pkg/schema/definitions/converter.go

```diff
@@ -0,0 +1,223 @@
+package definitions
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/rancher/apiserver/pkg/types"
+	wapiextv1 "github.com/rancher/wrangler/v2/pkg/generated/controllers/apiextensions.k8s.io/v1"
+	wranglerDefinition "github.com/rancher/wrangler/v2/pkg/schemas/definition"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/kube-openapi/pkg/util/proto"
+)
+
+var (
+	ErrNotFound     = errors.New("not found")
+	ErrNotRefreshed = errors.New("not refreshed")
+)
+
+// crdToDefinition builds a schemaDefinition for a CustomResourceDefinition
+func crdToDefinition(crdCache wapiextv1.CustomResourceDefinitionCache, crdName string, modelName string, version string) (schemaDefinition, error) {
+	crd, err := crdCache.Get(crdName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return schemaDefinition{}, ErrNotFound
+		}
+		return schemaDefinition{}, err
+	}
+
+	var jsonSchemaProps *apiextv1.JSONSchemaProps
+	for _, crdVersion := range crd.Spec.Versions {
+		if crdVersion.Name == version {
+			jsonSchemaProps = crdVersion.Schema.OpenAPIV3Schema
+			break
+		}
+	}
+
+	if jsonSchemaProps == nil {
+		return schemaDefinition{}, fmt.Errorf("unknown version %q for CRD %q", version, crdName)
+	}
+
+	// CRD definitions generally has more information than the OpenAPI V2
+	// because it embeds an OpenAPI V3 document. However, these 3 fields
+	// are the exception where the Open API V2 endpoint has more
+	// information.
+	//
+	// To avoid overriding these later on, we remove them here. Yeah, really.
+	delete(jsonSchemaProps.Properties, "apiVersion")
+	delete(jsonSchemaProps.Properties, "kind")
+	delete(jsonSchemaProps.Properties, "metadata")
+
+	path := proto.NewPath(modelName)
+
+	definitions := make(map[string]definition)
+	convertJSONSchemaPropsToDefinition(*jsonSchemaProps, path, definitions)
+
+	return schemaDefinition{
+		DefinitionType: modelName,
+		Definitions:    definitions,
+	}, nil
+}
+
+func convertJSONSchemaPropsToDefinition(props apiextv1.JSONSchemaProps, path proto.Path, definitions map[string]definition) {
+	if props.Type != "array" && props.Type != "object" {
```

Comment:
Here and elsewhere we seem to only cover the following types:
- array
- object
- string
- boolean
- integer
- number
How do we know that these are the only types?

Reply:
These are the only types defined by the OpenAPI V3 spec. I have added a comment that describes the available types and where they come from.

```go
// This supports all OpenAPI V3 types: boolean, number, integer, string, object and array
// as defined here: https://swagger.io/specification/v3/
```